src/transport/ExchangeMgr.h

```diff
@@ -0,0 +1,367 @@
+     * This function is the application callback to invoke when the timeout for the receipt
+     * of a response message has expired.
+     *
+     *  @param[in]    ec            A pointer to the ExchangeContext object.
+     *
+     */
+    typedef void (*ResponseTimeoutFunct)(ExchangeContext * ec);
+    ResponseTimeoutFunct OnResponseTimeout;
+
+    /**
+     * This function is the application callback to invoke when the timeout for the retransmission
+     * of a previously sent message has expired.
+     *
+     *  @param[in]    ec            A pointer to the ExchangeContext object.
+     *
+     */
+    typedef void (*RetransmissionTimeoutFunct)(ExchangeContext * ec);
+    RetransmissionTimeoutFunct OnRetransmissionTimeout;
+
+    /**
+     * This function is the application callback to invoke when an existing CHIP connection
+     * has been closed.
+     *
+     *  @param[in]    ec            A pointer to the ExchangeContext object.
+     *
+     *  @param[in]    con           A pointer to the PeerConnectionState object.
+     *
+     *  @param[in]    conErr        The CHIP_ERROR type that was reported when the connection
+     *                              was closed.
+     *
+     */
+    typedef void (*ConnectionClosedFunct)(ExchangeContext * ec, PeerConnectionState * con, CHIP_ERROR conErr);
+    ConnectionClosedFunct OnConnectionClosed;
+
+    /**
+     * Type of key error message handling function.
+     *
+     *  @param[in]    ec            A pointer to the ExchangeContext object.
+     *
+     *  @param[in]    keyErr        The CHIP_ERROR type that was reported in the key error message.
+     *
+     */
+    typedef void (*KeyErrorFunct)(ExchangeContext * ec, CHIP_ERROR keyErr);
+
+    /**
+     * This function is the application callback to invoke when key error message has been received
+     * from the peer.
+     */
+    KeyErrorFunct OnKeyError;
+
+    void CancelRetrans();
+
+    /*
+     * in order to use reference counting (see refCount below)
+     * we use a hold/free paradigm where users of the exchange
+     * can hold onto it while it's out of their direct control
+     * to make sure it isn't closed before everyone's ready.
+     */
+    void AddRef();
+    void Close();
+    void Abort();
+    void Release();
+    CHIP_ERROR StartTimerT();
+
+    enum
+    {
+        kGetPeerDescription_MaxLength = 100,
+        /**< Maximum length of string (including NUL character)*/
+    };
+
+    void GetPeerDescription(char * buf, uint32_t bufSize) const;
+
+private:
+    System::PacketBuffer * msg; // If we are re-transmitting, then this is the pointer to the message being retransmitted
+
+    // Trickle-controlled retransmissions:
+    uint32_t backoff;             // backoff for sampling the numner of messages
+    uint8_t msgsReceived;         // number of messages heard during the backoff period
+    uint8_t rebroadcastThreshold; // re-broadcast threshold
+    uint16_t mFlags;              // Internal state flags
```

Comment:
use chip::BitFlags to be typesafe (in support/BitFlags.h I believe)

Reply:
Ack